src/permission-set.js

```diff
@@ -16,523 +16,809 @@
+    var authList = []
+    var auth
+    var self = this
+    Object.keys(this.authorizations).forEach(function (authKey) {
+      auth = self.authorizations[ authKey ]
+      authList.push(auth)
+    })
+    return authList
   }
-  if (!accessMode) {
-    throw new Error('addPermission() requires a valid accessMode')
+
+  /**
+   * Tests whether this PermissionSet gives Public (acl:agentClass foaf:Agent)
+   * access to a given uri.
+   * @method allowsPublic
+   * @param mode {String|NamedNode} Access mode (read/write/control etc)
+   * @param resourceUrl {String}
+   * @return {Boolean}
+   */
+  allowsPublic (mode, resourceUrl) {
+    resourceUrl = resourceUrl || this.resourceUrl
+    let publicAuth = this.findPublicAuth(resourceUrl)
+    if (!publicAuth) {
+      return false
+    }
+    return this.checkOrigin(publicAuth) && publicAuth.allowsMode(mode)
   }
-  var auth = new Authorization(this.resourceUrl, this.isAuthInherited())
-  auth.setAgent(webId)
-  auth.addMode(accessMode)
-  if (origin) {
-    auth.addOrigin(origin)
+
+  /**
+   * Returns an RDF graph representation of this permission set and all its
+   * Authorizations. Used by `save()`.
+   * @method buildGraph
+   * @private
+   * @param rdf {RDF} RDF Library
+   * @return {Graph}
+   */
+  buildGraph (rdf) {
+    var graph = rdf.graph()
+    this.allAuthorizations().forEach(function (auth) {
+      graph.add(auth.rdfStatements(rdf))
+    })
+    return graph
   }
-  this.addAuthorization(auth)
-  return this
-}
-PermissionSet.prototype.addPermission = addPermission
 
-/**
- * Returns a list of all the Authorizations that belong to this permission set.
- * Mostly for internal use.
- * @method allAuthorizations
- * @return {Array<Authorization>}
- */
-function allAuthorizations () {
-  var authList = []
-  var auth
-  var self = this
-  Object.keys(this.authorizations).forEach(function (authKey) {
-    auth = self.authorizations[authKey]
-    authList.push(auth)
-  })
-  return authList
-}
-PermissionSet.prototype.allAuthorizations = allAuthorizations
+  /**
+   * Tests whether the given agent has the specified access to a resource.
+   * This is one of the main use cases for this solid-permissions library.
+   * Optionally performs strict origin checking (if `strictOrigin` is enabled
+   * in the constructor's options).
+   * Returns a promise; async since checking permissions may involve requesting
+   * multiple ACL resources (group listings, etc).
+   * @method checkAccess
+   * @param resourceUrl {String}
+   * @param agentId {String}
+   * @param accessMode {String} Access mode (read/write/control)
+   * @return {Promise<Boolean>}
+   */
+  checkAccess (resourceUrl, agentId, accessMode) {
+    let result
+    if (this.allowsPublic(accessMode, resourceUrl)) {
+      result = true
+    } else {
+      let auth = this.findAuthByAgent(agentId, resourceUrl)
+      result = auth && this.checkOrigin(auth) && auth.allowsMode(accessMode)
+    }
+    return Promise.resolve(result)
```

Comment:
The async part is coming in the next PR. (Has to do with fetching remote documents (like group listings)).